File: Bingo/Bingo.Core/StripBuilder.cs
namespace Bingo.Core
{
    class StripBuilder
    {
        public TicketBuilder[] Tickets { get; set; } = new TicketBuilder[6];

        public StripBuilder()
        {
            for (int i = 0; i < 6; i++)
            {
                Tickets[i] = new TicketBuilder(i);
            }
        }

        internal void Distribute(int v, int rangeIndex)
        {
            var random = new Random();
            var min = Tickets.Select(x => x.NumbersCount).Min();
            var a = new List<int>();
            var b = new List<int>();

            foreach (var tick in Tickets)
            {
                if (tick.NumbersCount == min)
                {
                    a.Add(tick.Index);
                }
                else
                {
                    b.Add(tick.Index);
                }
            }

            random.Shuffle1(a);
            var ordered = a.Concat(b).ToList();


            for (int i = 0; i < 6; i++)
            {
                var ticket = Tickets[ordered[i]];

                if (ticket.CanAccept(v, rangeIndex))
                {
                    ticket.Place(v, rangeIndex);
                    return;
                }
            }

            throw new Exception("This shouldn't happen");
        }

        internal Strip Build()
        {
            var tickets = Tickets.Select(t => t.Build()).ToArray();


            return new Strip(tickets);
        }
    }

    static class RandomExtensions
    {
        public static void Shuffle1(this Random rng, List<int> array)
        {
            int n = array.Count;
            while (n > 1)
            {
                int k = rng.Next(n--);
                int temp = array[n];
                array[n] = array[k];
                array[k] = temp;
            }
        }
    }

}

File: core/src/main/kotlin/org/example/core/generation/StripBuilder.kt
package org.example.core.generation

import org.example.core.Strip

class StripBuilder() {
    fun build(): Strip {
        return Strip(tickets.map { t -> t.build() }.toTypedArray())
    }

    fun distribute(value: Int, rangeIndex: Int) {
        val min = tickets.map { t -> t.numbersCount }.min()
        val a = mutableListOf<Int>()
        val b = mutableListOf<Int>()

        for (ticket in tickets){
            if (ticket.numbersCount == min) {
                a.add(ticket.index)
            } else{
                b.add(ticket.index)
            }
        }

        a.shuffle()
        val ordered = a + b

        for (i in tickets.indices){
            val ticket = tickets[ordered[i]]

            if (ticket.canAccept(value, rangeIndex)){
                ticket.place(value, rangeIndex)
                return
            }
        }
    }

    val tickets = Array(6) { TicketBuilder(it) }
}